### PR TITLE
nvproxy: expose CLOCK_MONOTONIC_RAW as a distinct clock for capProfiling

### PR DESCRIPTION
**Pull request by luiscape**
GPU profilers ([Nsight Systems](https://developer.nvidia.com/nsight-systems)/[CUPTI](https://developer.nvidia.com/cupti)) build their trace timeline from CLOCK_MONOTONIC_RAW. gVisor serves CLOCK_MONOTONIC_RAW as an alias of CLOCK_MONOTONIC, this will drift from CLOCK_MONOTONIC_RAW breaking profiling.

When nvproxy is enabled with `capProfiling`, this change exposes CLOCK_MONOTONIC_RAW as a clock distinct from CLOCK_MONOTONIC that tracks the host's CLOCK_MONOTONIC_RAW (absolute, unadjusted). CLOCK_MONOTONIC is unchanged. When profiling is disabled, CLOCK_MONOTONIC_RAW continues to alias CLOCK_MONOTONIC as before.

This is an example program that won't store kernel traces without this change:

```shell
  nsys profile --trace=cuda --sample=none --cpuctxsw=none \
    --export=sqlite -o /out/torch \
    python3 -c 'import torch; a=torch.randn(1024,1024,device="cuda"); (a@a).sum().item()'
```
One can verify that's the case by looking at the resulting SQLite database:

```shell
sqlite3 /tmp/torch.sqlite "SELECT count(*) FROM CUPTI_ACTIVITY_KIND_KERNEL"
```

Assisted-by: Claude

### CODE REVIEW

File: pkg/sentry/kernel/timekeeper.go [hunk 1]
```diff
@@ -69,6 +69,18 @@ type Timekeeper struct {
 	// It is set only once, by SetClocks.
 	monotonicOffset int64 `state:"nosave"`
```
Comment (aaltinaydev) [inline pkg/sentry/kernel/timekeeper.go:77]: In the commit message, "If the restore host's clock is behind the
saved monotonic time, monotonicity wins and alignment is abandoned"

Should we add a warning for this so it is not silent?

File: pkg/sentry/kernel/timekeeper.go [hunk 2]
```diff
@@ -134,6 +146,14 @@ func NewTimekeeper() *Timekeeper {
 	return &t
 }
 
+// SetHostAlignedMonotonic configures whether the monotonic clock is aligned
+// with the host's monotonic clock source rather than starting at zero on
+// boot, as required by GPU profiling tools. It must be called before
```
Comment (aaltinaydev) [inline pkg/sentry/kernel/timekeeper.go]: "must be called before SetClocks" but nothing enforces it.

Maybe add a similar check as in SetClocks: (if t.clocks != nil { panic(...) }) or pass a flag to the SetClocks and drop the setter.

**Review by aaltinaydev — DISMISSED** [review]

**Comment by luiscape** [comment]
@aaltinaydev thank you for the review. I added a warning and panic.

**Comment by ayushr2** [comment]
We may be iterating on this PR. Not ready to pull yet!

File: pkg/sentry/kernel/kernel.go [hunk 3]
```diff
@@ -1906,6 +1906,15 @@ func (k *Kernel) MonotonicClock() ktime.SampledClock {
 	return k.timekeeper.monotonicClock
 }
 
+// MonotonicRawClock returns the system CLOCK_MONOTONIC_RAW clock. When GPU
+// profiling is not enabled this is the same as MonotonicClock.
```
Comment (ayushr2) [inline pkg/sentry/kernel/kernel.go]: nit: The detail GPU profiling requires CLOCK_MONOTONIC_RAW is mentioned through many comments across the files. I think we should keep that detail in one place, i.e. where we construct the timekeeper with CLOCK_MONOTONIC_RAW support. Because in the future we can have more configuration which could we triggering CLOCK_MONOTONIC_RAW support. So it would be easier to update comments. And keeps this implementation independent of the GPU profiling details. I will add "nit: here too" comments below to refer to places where we can remove this context from the comment.

File: pkg/sentry/kernel/timekeeper.go [hunk 4]
```diff
@@ -69,6 +72,14 @@ type Timekeeper struct {
 	// It is set only once, by SetClocks.
 	monotonicOffset int64 `state:"nosave"`
 
+	// monotonicRawEnabled, if true, exposes CLOCK_MONOTONIC_RAW as a distinct
+	// clock that tracks the host's CLOCK_MONOTONIC_RAW; otherwise it aliases
+	// CLOCK_MONOTONIC.
+	//
+	// This supports GPU profiling (nvproxy CapProfiling) because profiling
+	// applications rely on CLOCK_MONOTONIC_RAW to anchor GPU timestamps.
```
Comment (ayushr2) [inline pkg/sentry/kernel/timekeeper.go]: nit: here too

File: vdso/vdso_time.cc [hunk 5]
```diff
@@ -41,6 +41,8 @@ struct params {
   int64_t realtime_base_cycles;
   int64_t realtime_base_ref;
   uint64_t realtime_frequency;
+
+  uint64_t monotonic_raw_syscall;
```
Comment (ayushr2) [inline vdso/vdso_time.cc]: nit: monotonic_raw_enabled?

File: pkg/sentry/kernel/timekeeper.go [hunk 6]
```diff
@@ -69,6 +72,14 @@ type Timekeeper struct {
...
+	monotonicRawEnabled bool
```
Comment (ayushr2) [inline pkg/sentry/kernel/timekeeper.go]: Actually, this field is not required. We can use `monotonicRawClock == nil` as the indicator that CLOCK_MONOTONIC_RAW support is not enabled. Consequently, NewTimekeeper() needs to take a parameter indicating if monotonic raw is enabled (which we can set based on sandbox configuration). We can delete SetMonotonicRawEnabled() below.

File: pkg/sentry/time/calibrated_clock.go [hunk 7]
```diff
@@ -244,10 +248,26 @@ func NewCalibratedClocks() *CalibratedClocks {
 	}
 }
 
+// NewProfilingCalibratedClocks creates a CalibratedClocks that additionally
+// tracks host CLOCK_MONOTONIC_RAW as a distinct clock (addressable as
+// MonotonicRaw). GPU profilers anchor the GPU timeline in the host
+// CLOCK_MONOTONIC_RAW domain, which drifts from CLOCK_MONOTONIC by NTP
+// frequency adjustment, so they require a sandbox clock in that domain.
```
Comment (ayushr2) [inline pkg/sentry/time/calibrated_clock.go]: nit: same here

File: runsc/boot/loader.go [hunk 8]
```diff
@@ -678,8 +684,13 @@ func New(args Args) (*Loader, error) {
 
 	// Create timekeeper.
 	tk := kernel.NewTimekeeper()
+	clocks := time.NewCalibratedClocks()
+	if monotonicRawForProfiling(args.Spec, args.Conf) {
+		clocks = time.NewProfilingCalibratedClocks()
```
Comment (ayushr2) [inline runsc/boot/loader.go]: Let's just add a parameter in NewCalibratedClocks() for monotonicRawEnabled?  Lesser code and it also avoids allocating a `clocks` object if monotonicRawForProfiling() returns true.

File: runsc/boot/loader.go [hunk 9]
```diff
@@ -515,6 +514,16 @@ func getRootCredentials(spec *specs.Spec, conf *config.Config, userNs *auth.User
 	return creds
 }
 
+// monotonicRawForProfiling reports whether CLOCK_MONOTONIC_RAW should be
+// exposed as a distinct clock tracking the host's CLOCK_MONOTONIC_RAW.
+func monotonicRawForProfiling(spec *specs.Spec, conf *config.Config) bool {
```
Comment (ayushr2) [inline runsc/boot/loader.go]: nit: maybe rename to `shouldEnableClockMonotonicRaw()`. It's implementation in the future may be expanded to allow it more broadly.

File: pkg/sentry/time/calibrated_clock.go [hunk 10]
```diff
@@ -259,6 +279,12 @@ func (c *CalibratedClocks) GetTime(id ClockID) (int64, error) {
 		return c.monotonic.GetTime()
 	case Realtime:
 		return c.realtime.GetTime()
+	case MonotonicRaw:
+		if c.monotonicRaw != nil {
+			return c.monotonicRaw.GetTime()
```
Comment (ayushr2) [inline pkg/sentry/time/calibrated_clock.go:283]: Hmm, does this make the host syscall to fetch the host CLOCK_MONOTONIC_RAW value?

File: pkg/sentry/time/clock_id.go [hunk 11]
```diff
@@ -23,8 +23,9 @@ type ClockID int32
 
 // These are the supported Linux clock identifiers.
 const (
-	Realtime ClockID = iota
-	Monotonic
+	Realtime     ClockID = 0
+	Monotonic    ClockID = 1
+	MonotonicRaw ClockID = 4
```
Comment (ayushr2) [inline pkg/sentry/time/clock_id.go:28]: Wow, good catch. That was quite the footgun to define this as iota. The numeric value of the ClockID matters since it is passed into `vdsoClockGettime()` in pkg/sentry/time/sampler.go.

**Review by ayushr2 — APPROVED** [review]
Thanks Luis!! Just nits.

File: pkg/sentry/time/calibrated_clock.go (hunk 10 above)
Comment (ayushr2) [reply pkg/sentry/time/calibrated_clock.go:283]: Yes, it does.

File: vdso/vdso_time.cc [hunk 12]
```diff
@@ -156,4 +158,13 @@ int ClockMonotonic(struct timespec* ts) {
   return 0;
 }
 
+// ClockMonotonicRaw() is the VDSO implementation of
+// clock_gettime(CLOCK_MONOTONIC_RAW).
+int ClockMonotonicRaw(struct timespec* ts) {
+  if (get_params()->monotonic_raw_syscall) {
+    return sys_clock_gettime(CLOCK_MONOTONIC_RAW, ts);
```
Comment (ayushr2) [inline vdso/vdso_time.cc:203]: Profilers timestamp at high rates inside the process being measured, so a syscall per timestamp will probably hurt and give poor timestamp measurements. VDSO exists to avoid the syscall context switch.

Can we serve this from the VDSO instead of falling back to the syscall? The param page has plenty of room (currently seq + 8 u64s in a 4KiB page), and all clocks calibrate against the same TSC, so it's mostly mechanical:
- add monotonicRaw{Ready,BaseCycles,BaseRef,Frequency} to vdsoParams and mirror them in struct params
- publish them in Timekeeper.update(), without monotonicOffset
- make ClockMonotonicRaw() a copy of ClockMonotonic() reading the raw fields

Clocks.Update() would need to return the raw params too. Six return values is a lot; maybe fold them into a struct.

One thing to be careful about: we need to distinguish "raw disabled" from "raw enabled but not yet calibrated". If we only check ready and fall back to the syscall, every non-profiling sandbox regresses from today's free VDSO alias to a syscall. So !enabled should call ClockMonotonic() directly, and only enabled && !ready should hit the syscall:
```
int ClockMonotonicRaw(struct timespec* ts) {
  // ...seqcount loop reading enabled, ready, base_ref, base_cycles, frequency...
  if (!enabled) return ClockMonotonic(ts);                    // in-VDSO alias, as today
  if (!ready)   return sys_clock_gettime(CLOCK_MONOTONIC_RAW, ts);
  // ...same delta_cycles/cycles_to_ns computation...
}
```

I think accuracy should not be impacted, since the syscall path uses the same calibrated parameters the VDSO would recompute.

Comment (luiscape) [reply vdso/vdso_time.cc:203]: @ayushr2 `ClockMonotonicRaw()` now keeps track of `CLOCK_MONOTONIC_RAW` clock and returns the VDSO version instead of making a syscall--much better approach, thank you. It calibrates the VDSO raw clock during first container start-up and restore, but just returns VDSO version from there onwards.

File: vdso/vdso_time.cc [hunk 13]
```diff
@@ -149,10 +165,49 @@ int ClockMonotonic(struct timespec* ts) {
...
+    seq = read_seqcount_begin(&params->seq_count);
+    enabled = params->monotonic_raw_enabled;
+    ready = params->monotonic_raw_ready;
+    base_ref = params->monotonic_raw_base_ref;
+    base_cycles = params->monotonic_raw_base_cycles;
+    frequency = params->monotonic_raw_frequency;
+    now_cycles = cycle_clock();
+  } while (read_seqcount_retry(&params->seq_count, seq));
+
+  if (!enabled) {
```
Comment (ayushr2) [inline vdso/vdso_time.cc]: I think a zeroed param page reaching this branch is a problem. On restore, `SetClocks()` explicitly resets the params to `vdsoParams{}`, and `startUpdater()` republishes them asynchronously from a goroutine while `close(t.restored)` happens right away. The VDSO path doesn't synchronize on `t.restored`, so a restored app thread can read the page before the first `update()`.

In that window `monotonic_raw_enabled == 0`, so we take the alias branch into `ClockMonotonic()`, and since `monotonic_ready` is also 0 there, we end up returning CLOCK_MONOTONIC. The app asked for CLOCK_MONOTONIC_RAW and silently gets a value off by the host's uptime -- which is the exact failure this change is meant to fix.

The issue is that all-zeros currently means "alias" when it should mean "sentry hasn't told us yet". Can we invert the polarity? Something like `monotonic_raw_alias`, set to 1 in `update()` when `t.monotonicRawClock == nil`:

```c
if (alias)  return ClockMonotonic(ts);
if (!ready) return sys_clock_gettime(CLOCK_MONOTONIC_RAW, ts);
```

A zeroed page then falls through to the syscall and the sentry resolves it correctly in either configuration. Non-profiling sandboxes take the syscall during the same brief window, which is what monotonic and realtime already do when `!ready`.

File: pkg/sentry/syscalls/linux/sys_time.go [hunk 14]
```diff
@@ -123,9 +123,10 @@ func getClock(t *kernel.Task, clockID int32) (ktime.Clock, error) {
 	switch clockID {
 	case linux.CLOCK_REALTIME, linux.CLOCK_REALTIME_COARSE:
 		return t.Kernel().RealtimeClock(), nil
+	case linux.CLOCK_MONOTONIC_RAW:
```
Comment (ayushr2) [inline pkg/sentry/syscalls/linux/sys_time.go:126]: `getClock()` also backs `ClockGetres`, `ClockNanosleep` and `timer_create`. Linux's `clock_monotonic_raw` k_clock only defines `clock_getres` and `clock_get_timespec`, so `timer_create(CLOCK_MONOTONIC_RAW)` returns EOPNOTSUPP and `clock_nanosleep(CLOCK_MONOTONIC_RAW)` returns ENOTSUP.

We accepted both before by aliasing to CLOCK_MONOTONIC, which was wrong but harmless. Now the clock is in a different epoch, so a TIMER_ABSTIME sleep or timer gets interpreted against host boot time. It also interacts badly with C/R: `monotonicRawClock` is `state:"nosave"`, but a saved `IntervalTimer` can still hold a `*timekeeperClock{c: MonotonicRaw}` through its own reference. If such a timer is restored into a sandbox with the opposite configuration, its absolute expiration is off by roughly the host uptime in one direction or the other, so it either fires immediately or effectively never.

Can we reject CLOCK_MONOTONIC_RAW for `timer_create` and `clock_nanosleep`? That matches Linux and makes the C/R case moot.

File: pkg/sentry/kernel/timekeeper.go [hunk 15]
```diff
@@ -158,6 +164,14 @@ func (t *Timekeeper) SetClocks(c sentrytime.Clocks, params *VDSOParamPage) {
 
 	t.clocks = c
 
+	// Serve CLOCK_MONOTONIC_RAW as a distinct clock iff the clock source
+	// tracks it (see NewCalibratedClocks). Deriving this from the source keeps
+	// boot and restore coherent: a restored sandbox follows its current
+	// configuration, not the checkpointed one.
+	if _, err := c.GetTime(sentrytime.MonotonicRaw); err == nil {
```
Comment (ayushr2) [inline pkg/sentry/kernel/timekeeper.go]: Deriving this from the clock source rather than the checkpoint is the right call, but I'd rather not detect it by probing for an error. This works today only because `CalibratedClock.GetTime()` falls back to `c.ref.Syscall()` when the clock isn't ready yet, and `syscallTSCReferenceClocks.Sample()` happens to never return a non-nil error -- so an uncalibrated raw clock still probes as supported. But `sampler.Syscall()` does propagate errors, so if that ever changes, CLOCK_MONOTONIC_RAW silently degrades to an alias for the sandbox's entire lifetime with nothing in the logs.

Could we add a `MonotonicRawEnabled() bool` to the `Clocks` interface, or just pass the bool to `SetClocks()`? Either states the intent directly and doesn't depend on that fallback behavior.

File: vdso/vdso_time.cc (hunk 13 above)
Comment (luiscape) [reply vdso/vdso_time.cc]: Makes sense. Just updated so now a all-zero page will falls through to the syscall so the sentry resolves `CLOCK_MONOTONIC_RAW` authoritatively during the restore window.

File: pkg/sentry/syscalls/linux/sys_time.go (hunk 14 above)
Comment (luiscape) [reply pkg/sentry/syscalls/linux/sys_time.go:126]: `timer_create` now returns EOPNOTSUPP.

**Review by ayushr2 — APPROVED** [review]
Thanks!

**Review by aaltinaydev — APPROVED** [review]
